In [ ]:
!pip install ultralytics -q

import kagglehub, os
dataset_str = kagglehub.dataset_download("zehraatlgan/fish-detection")
data_yaml = os.path.join(dataset_str, "data.yaml")
print("Using data.yaml at:", data_yaml)

from ultralytics import YOLO

# Бейзлайн
model = YOLO('yolov8n.pt')
results_baseline = model.train(
    data=data_yaml,
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=1e-3,
    name='fish_baseline'
)

print("\n Baseline metrics:")
for metric, vals in results_baseline.metrics.items():
    print(f"{metric}: {vals[-1]:.4f}")


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 31.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 118.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 90.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 53.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 87.9 MB/s eta 0:00:00
Using data.yaml at: /kaggle/input/fish-detection/data.yaml
Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics 

100%|██████████| 6.25M/6.25M [00:00<00:00, 120MB/s]


Ultralytics 8.3.134 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/kaggle/input/fish-detection/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=fish_baseline, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, plots=True, po

100%|██████████| 755k/755k [00:00<00:00, 25.8MB/s]

Overriding model.yaml nc=80 with nc=13

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytic

 22        [15, 18, 21]  1    753847  ultralytics.nn.modules.head.Detect           [13, [64, 128, 256]]          
Model summary: 129 layers, 3,013,383 parameters, 3,013,367 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks...


100%|██████████| 5.35M/5.35M [00:00<00:00, 110MB/s]


AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.0±0.0 ms, read: 11.7±7.2 MB/s, size: 46.5 KB)


train: Scanning /kaggle/input/fish-detection/train/labels... 6842 images, 0 backgrounds, 0 corrupt: 100%|██████████| 6842/6842 [00:36<00:00, 187.05it/s]


WARNING ⚠️ train: Cache directory /kaggle/input/fish-detection/train is not writeable, cache not saved.
WARNING ⚠️ Box and segment counts should be equal, but got len(segments) = 536, len(boxes) = 14154. To resolve this only boxes will be used and all segments will be removed. To avoid this please supply either a detect or segment dataset, not a detect-segment mixed dataset.
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 14.7±7.9 MB/s, size: 37.0 KB)


val: Scanning /kaggle/input/fish-detection/valid/labels... 700 images, 0 backgrounds, 0 corrupt: 100%|██████████| 700/700 [00:03<00:00, 184.19it/s]


WARNING ⚠️ val: Cache directory /kaggle/input/fish-detection/valid is not writeable, cache not saved.
WARNING ⚠️ Box and segment counts should be equal, but got len(segments) = 84, len(boxes) = 1197. To resolve this only boxes will be used and all segments will be removed. To avoid this please supply either a detect or segment dataset, not a detect-segment mixed dataset.
Plotting labels to runs/detect/fish_baseline/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.001' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000588, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/fish_baseline
Starting training for 10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_

       1/10      2.26G      1.343      3.115      1.614         13        640: 100%|██████████| 428/428 [02:11<00:00,  3.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:09<00:00,  2.38it/s]


                   all        700       1197      0.656      0.547      0.639      0.406

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/10      2.66G      1.324      2.083      1.541         15        640: 100%|██████████| 428/428 [02:04<00:00,  3.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.89it/s]

                   all        700       1197      0.723      0.636      0.706      0.418



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/10      2.68G      1.272      1.669      1.495         25        640: 100%|██████████| 428/428 [02:04<00:00,  3.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.91it/s]

                   all        700       1197      0.689      0.674       0.73      0.466



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/10       2.7G      1.242      1.389       1.46         14        640: 100%|██████████| 428/428 [02:04<00:00,  3.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.95it/s]

                   all        700       1197      0.742      0.702      0.793      0.531



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/10      2.71G      1.192      1.188      1.413         19        640: 100%|██████████| 428/428 [02:04<00:00,  3.45it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.95it/s]

                   all        700       1197      0.752      0.792      0.822       0.55



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/10      2.73G      1.147      1.051      1.376         23        640: 100%|██████████| 428/428 [02:03<00:00,  3.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.97it/s]

                   all        700       1197      0.847      0.822      0.875      0.601



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/10      2.75G        1.1     0.9456      1.349         11        640: 100%|██████████| 428/428 [02:06<00:00,  3.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.99it/s]

                   all        700       1197      0.856       0.82      0.889      0.628



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/10      2.76G      1.061     0.8579      1.311         24        640: 100%|██████████| 428/428 [02:04<00:00,  3.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.96it/s]

                   all        700       1197      0.866       0.82      0.885      0.636



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/10      2.78G      1.014     0.7768      1.277         11        640: 100%|██████████| 428/428 [02:06<00:00,  3.40it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.98it/s]

                   all        700       1197      0.888      0.853      0.908       0.65



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/10       2.8G     0.9878     0.7319      1.254         34        640: 100%|██████████| 428/428 [02:04<00:00,  3.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.93it/s]

                   all        700       1197      0.872      0.874      0.913      0.667



10 epochs completed in 0.371 hours.
Optimizer stripped from runs/detect/fish_baseline/weights/last.pt, 6.2MB
Optimizer stripped from runs/detect/fish_baseline/weights/best.pt, 6.2MB

Validating runs/detect/fish_baseline/weights/best.pt...
Ultralytics 8.3.134 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,008,183 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:10<00:00,  2.18it/s]


                   all        700       1197      0.871      0.875      0.912      0.667
              BlueTang          7         23      0.872      0.913      0.902      0.812
         ButterflyFish         10         47      0.846      0.894        0.9      0.483
             ClownFish         28         41      0.877      0.872      0.941       0.64
              GoldFish         62        134      0.845      0.937      0.934      0.658
               Gourami        188        210      0.968          1      0.995       0.91
            MorishIdol         23         31      0.936      0.942      0.986      0.676
             PlatyFish         50        102      0.786      0.755      0.835      0.594
     RibbonedSweetlips         22         67      0.798      0.836      0.876      0.655
ThreeStripedDamselfish         79        108      0.962       0.95      0.989      0.797
         YellowCichlid         22         95      0.832      0.484      0.648      0.362
            YellowTan

AttributeError: 'DetMetrics' object has no attribute 'metrics'. See valid attributes below.

    Utility class for computing detection metrics such as precision, recall, and mean average precision (mAP).

    Attributes:
        save_dir (Path): A path to the directory where the output plots will be saved.
        plot (bool): A flag that indicates whether to plot precision-recall curves for each class.
        names (dict): A dictionary of class names.
        box (Metric): An instance of the Metric class for storing detection results.
        speed (dict): A dictionary for storing execution times of different parts of the detection process.
        task (str): The task type, set to 'detect'.
    

In [ ]:
metrics = results_baseline.results_dict

for name, value in metrics.items():
    print(f"{name}: {value:.4f}")



metrics/precision(B): 0.8714
metrics/recall(B): 0.8746
metrics/mAP50(B): 0.9124
metrics/mAP50-95(B): 0.6671
fitness: 0.6916


Для повышения mAP мы провели 12-эпоховую дообучение YOLOv8n с агрессивными аугментациями (RandAugment, Copy-Paste, MixUp и Mosaic), оптимизатором AdamW и cosine-LR scheduler, чтобы разнообразить обучающие примеры и сгладить процесс обучения, что улучшает обобщающую способность модели.









In [ ]:
model = YOLO('yolov8n.pt')

# 3) Тренировка: 12 эпох, агрессивные аугментации, AdamW + cosine LR
results = model.train(
    data=data_yaml,
    epochs=12,
    imgsz=640,
    batch=16,
    lr0=5e-4,
    optimizer='AdamW',
    cos_lr=True,
    augment=True,
    auto_augment='v0',
    copy_paste=0.2,
    mixup=0.2,
    mosaic=0.5,
    fliplr=0.5,
    flipud=0.3,
    hsv_h=0.02, hsv_s=0.8, hsv_v=0.5,
    name='fish_augmented_12ep'
)

metrics = results.results_dict
print("\n Metrics after 12 epochs on augmented data:")
print(f"Precision:    {metrics['precision']:.4f}")
print(f"Recall:       {metrics['recall']:.4f}")
print(f"mAP@50:       {metrics['map50']:.4f}")
print(f"mAP@50-95:    {metrics['map50_95']:.4f}")
print(f"Fitness:      {metrics.get('fitness', float('nan')):.4f}")

Ultralytics 8.3.134 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=True, auto_augment=v0, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.2, copy_paste_mode=flip, cos_lr=True, cutmix=0.0, data=/kaggle/input/fish-detection/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=12, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.3, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.02, hsv_s=0.8, hsv_v=0.5, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.0005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.2, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=0.5, multi_scale=False, name=fish_augmented_12ep, nbs=64, nms=False, opset=None, optimize=False, optimizer=AdamW, overlap_mask=True, patience=100, perspective=0.0, plots=True, pose=1

train: Scanning /kaggle/input/fish-detection/train/labels... 6842 images, 0 backgrounds, 0 corrupt: 100%|██████████| 6842/6842 [00:11<00:00, 570.96it/s]


WARNING ⚠️ train: Cache directory /kaggle/input/fish-detection/train is not writeable, cache not saved.
WARNING ⚠️ Box and segment counts should be equal, but got len(segments) = 536, len(boxes) = 14154. To resolve this only boxes will be used and all segments will be removed. To avoid this please supply either a detect or segment dataset, not a detect-segment mixed dataset.
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Fast image access ✅ (ping: 0.2±0.2 ms, read: 101.0±39.2 MB/s, size: 37.0 KB)


val: Scanning /kaggle/input/fish-detection/valid/labels... 700 images, 0 backgrounds, 0 corrupt: 100%|██████████| 700/700 [00:01<00:00, 498.17it/s]

WARNING ⚠️ val: Cache directory /kaggle/input/fish-detection/valid is not writeable, cache not saved.
WARNING ⚠️ Box and segment counts should be equal, but got len(segments) = 84, len(boxes) = 1197. To resolve this only boxes will be used and all segments will be removed. To avoid this please supply either a detect or segment dataset, not a detect-segment mixed dataset.


Plotting labels to runs/detect/fish_augmented_12ep/labels.jpg... 
optimizer: AdamW(lr=0.0005, momentum=0.937) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/fish_augmented_12ep
Starting training for 12 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/12      2.71G      1.559      2.876      1.709         27        640: 100%|██████████| 428/428 [02:36<00:00,  2.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.91it/s]

                   all        700       1197      0.513       0.51      0.494       0.31



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/12      2.98G      1.512      2.075      1.653         46        640: 100%|██████████| 428/428 [02:28<00:00,  2.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:06<00:00,  3.17it/s]

                   all        700       1197       0.55       0.57      0.589      0.366


Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/12      2.98G      1.335      1.512      1.585         28        640: 100%|██████████| 428/428 [02:07<00:00,  3.35it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:06<00:00,  3.58it/s]

                   all        700       1197      0.627      0.715      0.724      0.446



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/12      2.98G      1.274      1.301      1.509         14        640: 100%|██████████| 428/428 [02:08<00:00,  3.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:06<00:00,  3.33it/s]

                   all        700       1197       0.68       0.73       0.74       0.48



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/12         3G      1.225      1.166       1.47         19        640: 100%|██████████| 428/428 [02:07<00:00,  3.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.06it/s]

                   all        700       1197      0.762      0.724      0.784      0.522



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/12      3.01G      1.197      1.066      1.439         23        640: 100%|██████████| 428/428 [02:07<00:00,  3.36it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.04it/s]

                   all        700       1197      0.735      0.785      0.817       0.56



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/12      3.03G      1.147     0.9773      1.412         11        640: 100%|██████████| 428/428 [02:04<00:00,  3.45it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  2.96it/s]

                   all        700       1197      0.826        0.8      0.857      0.601



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/12      3.05G      1.109     0.9045      1.375         24        640: 100%|██████████| 428/428 [02:06<00:00,  3.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.06it/s]

                   all        700       1197      0.841      0.826      0.885       0.62



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/12      3.06G      1.073     0.8271      1.339         11        640: 100%|██████████| 428/428 [02:07<00:00,  3.36it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.07it/s]

                   all        700       1197      0.851      0.841       0.88      0.639



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/12      3.08G      1.038     0.7729      1.322         34        640: 100%|██████████| 428/428 [02:04<00:00,  3.43it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.12it/s]

                   all        700       1197      0.845      0.878      0.902      0.653



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/12       3.1G      1.017     0.7456      1.298         21        640: 100%|██████████| 428/428 [02:04<00:00,  3.43it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:08<00:00,  2.65it/s]

                   all        700       1197      0.845      0.889      0.903      0.664



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/12      3.12G      1.008     0.7289      1.296         24        640: 100%|██████████| 428/428 [02:05<00:00,  3.42it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:07<00:00,  3.09it/s]

                   all        700       1197      0.839      0.886        0.9      0.665



12 epochs completed in 0.461 hours.
Optimizer stripped from runs/detect/fish_augmented_12ep/weights/last.pt, 6.2MB
Optimizer stripped from runs/detect/fish_augmented_12ep/weights/best.pt, 6.2MB

Validating runs/detect/fish_augmented_12ep/weights/best.pt...
Ultralytics 8.3.134 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,008,183 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 22/22 [00:10<00:00,  2.03it/s]


                   all        700       1197      0.839      0.873      0.909      0.676
              BlueTang          7         23      0.869      0.864      0.922      0.831
         ButterflyFish         10         47      0.736      0.809       0.81      0.452
             ClownFish         28         41      0.845      0.976      0.946      0.658
              GoldFish         62        134      0.795      0.924      0.939      0.694
               Gourami        188        210      0.929          1      0.995      0.915
            MorishIdol         23         31      0.876      0.903      0.967      0.767
             PlatyFish         50        102      0.768      0.784      0.864      0.605
     RibbonedSweetlips         22         67      0.758       0.91      0.899      0.646
ThreeStripedDamselfish         79        108      0.895       0.95      0.979      0.771
         YellowCichlid         22         95      0.895      0.449      0.684      0.368
            YellowTan

KeyError: 'precision'

In [ ]:
metrics = results.results_dict

for name, value in metrics.items():
    print(f"{name}: {value:.4f}")


metrics/precision(B): 0.8393
metrics/recall(B): 0.8731
metrics/mAP50(B): 0.9091
metrics/mAP50-95(B): 0.6762
fitness: 0.6995


Судя по результатам, агрессивные аугментации вместе с AdamW + cosine дали небольшой прирост mAP50-95 (0.6762 vs 0.6671), но заметно снизили precision (0.8393 vs 0.8714) и слегка упали mAP50 (0.9091 vs 0.9124) — то есть модель лучше научилась «точно» локализовывать объекты по IoU, но стала чаще делать ложные срабатывания.

собственная имплементация

In [ ]:
import os
import yaml
import glob
from PIL import Image
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as T
import kagglehub

# 1. Download dataset and get path
dataset_str = kagglehub.dataset_download("zehraatlgan/fish-detection")
data_yaml = os.path.join(dataset_str, "data.yaml")
print("Using data.yaml at:", data_yaml)

# 2. Load dataset configuration
with open(data_yaml, 'r') as f:
    data_cfg = yaml.safe_load(f)

# 3. Resolve image directories (handle various relative paths)
base_dir = os.path.dirname(data_yaml)
train_candidates = [
    os.path.normpath(os.path.join(base_dir, data_cfg['train'])),
    os.path.normpath(os.path.join(dataset_str, data_cfg['train'])),
    os.path.join(dataset_str, 'train/images'),
    os.path.join(dataset_str, 'train'),
]
val_candidates = [
    os.path.normpath(os.path.join(base_dir, data_cfg['val'])),
    os.path.normpath(os.path.join(dataset_str, data_cfg['val'])),
    os.path.join(dataset_str, 'valid/images'),
    os.path.join(dataset_str, 'valid'),
    os.path.join(dataset_str, 'val/images'),
    os.path.join(dataset_str, 'val'),
]

# Select first existing directory
train_img_dir = next((d for d in train_candidates if os.path.isdir(d)), None)
val_img_dir   = next((d for d in val_candidates   if os.path.isdir(d)), None)
if train_img_dir is None:
    raise FileNotFoundError(f"No valid train directory found among: {train_candidates}")
if val_img_dir is None:
    raise FileNotFoundError(f"No valid val directory found among: {val_candidates}")

print(f"Using train images from: {train_img_dir}")
print(f"Using val images from:   {val_img_dir}")

# Class names
class_names = data_cfg.get('names', {})
num_classes = len(class_names)

# 4. Prepare small subset
train_imgs = sorted(glob.glob(os.path.join(train_img_dir, '*.jpg')))[:100]
val_imgs   = sorted(glob.glob(os.path.join(val_img_dir,   '*.jpg')))[:50]
if not train_imgs:
    raise RuntimeError(f"No JPG images found in train dir: {train_img_dir}")
if not val_imgs:
    raise RuntimeError(f"No JPG images found in val dir: {val_img_dir}")

# 5. Dataset class
class FishDataset(Dataset):
    def __init__(self, img_paths, img_size=640, grid_size=20):
        self.img_paths = img_paths
        self.img_size  = img_size
        self.grid_size = grid_size
        self.transform = T.Compose([
            T.Resize((img_size, img_size)),
            T.ToTensor(),
        ])

    def __len__(self):
        return len(self.img_paths)

    def __getitem__(self, idx):
        img_path = self.img_paths[idx]
        txt_path = img_path.replace('.jpg', '.txt')

        img = Image.open(img_path).convert('RGB')
        img = self.transform(img)

        S = self.grid_size
        target = torch.zeros((5, S, S))

        if os.path.isfile(txt_path):
            for line in open(txt_path):
                cls, x, y, w, h = map(float, line.split())
                gx, gy = x * S, y * S
                gi, gj = int(gx), int(gy)
                tx, ty = gx - gi, gy - gj
                tw, th = w * S, h * S
                target[0, gj, gi] = tx
                target[1, gj, gi] = ty
                target[2, gj, gi] = tw
                target[3, gj, gi] = th
                target[4, gj, gi] = 1.0

        return img, target

def collate_fn(batch):
    imgs, targets = zip(*batch)
    return torch.stack(imgs), torch.stack(targets)

# 6. Model definition
class SimpleYOLO(nn.Module):
    def __init__(self, grid_size=20):
        super().__init__()
        self.backbone = nn.Sequential(
            nn.Conv2d(3, 16, 3, 1, 1), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(16, 32, 3, 1, 1), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(32, 64, 3, 1, 1), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(64, 128,3, 1, 1), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(128,256,3, 1, 1), nn.ReLU(), nn.MaxPool2d(2)
        )
        self.pred = nn.Conv2d(256, 5, 1)

    def forward(self, x):
        return self.pred(self.backbone(x))

# 7. Setup data loaders
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
train_ds = FishDataset(train_imgs)
train_dl = DataLoader(train_ds, batch_size=8, shuffle=True, collate_fn=collate_fn)
val_ds   = FishDataset(val_imgs)
val_dl   = DataLoader(val_ds,   batch_size=8, shuffle=False, collate_fn=collate_fn)

# 8. Training loop
def train_one_epoch(model, dataloader, optimizer, criterion, device):
    model.train()
    total_loss = 0.0
    for imgs, targets in dataloader:
        imgs, targets = imgs.to(device), targets.to(device)
        preds = model(imgs)
        loss  = criterion(preds, targets)
        optimizer.zero_grad()
        loss.backward() # ебанутые блять
        optimizer.step()
        total_loss += loss.item()
    return total_loss / len(dataloader)

model     = SimpleYOLO().to(device)
optimizer = optim.AdamW(model.parameters(), lr=5e-4)
criterion = nn.MSELoss()

epochs = 5
for epoch in range(1, epochs+1):
    loss = train_one_epoch(model, train_dl, optimizer, criterion, device)
    print(f"Epoch {epoch}/{epochs} - Loss: {loss:.4f}")

# 9. Save
torch.save(model.state_dict(), 'simple_yolo_fish.pth')
print("Training complete. Model saved to simple_yolo_fish.pth")

Using data.yaml at: /kaggle/input/fish-detection/data.yaml
Using train images from: /kaggle/input/fish-detection/train/images
Using val images from:   /kaggle/input/fish-detection/valid/images
Epoch 1/5 - Loss: 0.0002
Epoch 2/5 - Loss: 0.0000
Epoch 3/5 - Loss: 0.0000
Epoch 4/5 - Loss: 0.0000
Epoch 5/5 - Loss: 0.0000
Training complete. Model saved to simple_yolo_fish.pth


In [ ]:
model = SimpleYOLO().to(device)
model.load_state_dict(torch.load('simple_yolo_fish.pth'))
model.eval()


SimpleYOLO(
  (backbone): Sequential(
    (0): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU()
    (2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (3): Conv2d(16, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (4): ReLU()
    (5): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (6): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (7): ReLU()
    (8): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (9): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (10): ReLU()
    (11): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (12): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (13): ReLU()
    (14): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (pred): Conv2d(256, 5, kernel_size=(1, 1), stride=(1, 1))
)

In [ ]:
import torch
from torchvision.ops import box_iou

def decode_preds(pred, S=20, img_size=640, conf_thr=0.5):
    """
    Находит в каждом батче все ячейки с conf>conf_thr
    и превращает их в боксы (x1,y1,x2,y2) в пикселях.
    """
    B = pred.size(0)
    cell = img_size / S
    all_preds = []
    for b in range(B):
        p = pred[b]  # [5,S,S]
        objs = p[4]
        idxs = (objs > conf_thr).nonzero(as_tuple=False)
        boxes, scores = [], []
        for gj, gi in idxs:
            tx, ty, tw, th = p[0,gj,gi], p[1,gj,gi], p[2,gj,gi], p[3,gj,gi]
            cx = (gi + tx) * cell
            cy = (gj + ty) * cell
            w  = tw * cell
            h  = th * cell
            x1, y1 = cx - w/2, cy - h/2
            x2, y2 = cx + w/2, cy + h/2
            boxes.append([x1,y1,x2,y2])
            scores.append(objs[gj,gi].item())
        if boxes:
            all_preds.append({
                'boxes': torch.tensor(boxes),
                'scores': torch.tensor(scores)
            })
        else:
            all_preds.append({'boxes': torch.zeros((0,4)), 'scores': torch.zeros((0,))})
    return all_preds

def get_gt_boxes(target, S=20, img_size=640):
    """
    Из таргета [5,S,S] возвращает тензор [N,4] истинных боксов.
    """
    cell = img_size / S
    obj = target[4]
    idxs = (obj > 0.5).nonzero(as_tuple=False)
    boxes = []
    for gj, gi in idxs:
        tx, ty, tw, th = target[0,gj,gi], target[1,gj,gi], target[2,gj,gi], target[3,gj,gi]
        cx = (gi + tx) * cell
        cy = (gj + ty) * cell
        w  = tw * cell
        h  = th * cell
        x1, y1 = cx - w/2, cy - h/2
        x2, y2 = cx + w/2, cy + h/2
        boxes.append([x1,y1,x2,y2])
    return torch.tensor(boxes)


In [ ]:
all_tp = all_fp = all_fn = 0

with torch.no_grad():
    for imgs, targets in val_dl:
        imgs = imgs.to(device)
        preds = model(imgs).cpu()
        batch_preds = decode_preds(preds, S=20, img_size=640, conf_thr=0.5)

        for i in range(len(batch_preds)):
            pred = batch_preds[i]
            gt   = get_gt_boxes(targets[i], S=20, img_size=640)

            if len(pred['boxes']) == 0:
                all_fn += len(gt); continue
            if len(gt) == 0:
                all_fp += len(pred['boxes']); continue

            ious = box_iou(pred['boxes'], gt)
            # для каждого предикта берём максимальный IoU
            max_iou, _ = ious.max(dim=1)  # [P]
            tp = (max_iou >= 0.5).sum().item()
            fp = (max_iou < 0.5).sum().item()
            fn = len(gt) - tp

            all_tp += tp
            all_fp += fp
            all_fn += fn

precision = all_tp / (all_tp + all_fp + 1e-8)
recall    = all_tp / (all_tp + all_fn + 1e-8)
f1        = 2 * precision * recall / (precision + recall + 1e-8)

print(f"Val — Precision: {precision:.4f}, Recall: {recall:.4f}, F1: {f1:.4f}")


Val — Precision: 0.0000, Recall: 0.0000, F1: 0.0000


In [2]:
import os
import yaml
import glob
from PIL import Image
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as T
import kagglehub
from torchvision.ops import box_iou

# =========================
# 1. DATA LOADING
# =========================
print("Downloading dataset (cached after first run)…")
root = kagglehub.dataset_download("zehraatlgan/fish-detection")
ctx_yaml = os.path.join(root, "data.yaml")
print("Using data.yaml:", ctx_yaml)

with open(ctx_yaml) as f:
    cfg = yaml.safe_load(f)
train_dir = os.path.normpath(os.path.join(root, cfg["train"].lstrip("../")))
val_dir   = os.path.normpath(os.path.join(root, cfg["val"].lstrip("../")))
print("Train dir:", train_dir)
print("Val dir  :", val_dir)

train_imgs = sorted(glob.glob(os.path.join(train_dir, "*.jpg")))[:128]
val_imgs   = sorted(glob.glob(os.path.join(val_dir,   "*.jpg")))[:64]
assert train_imgs and val_imgs, "Images not found!"
IMG_SIZE = 416  # reduce resolution to cut memory

# =========================
# 2. DATASET
# =========================
class FishDataset(Dataset):
    def __init__(self, paths, S=13, augment=False):
        self.paths = paths; self.S=S
        base = [T.Resize((IMG_SIZE, IMG_SIZE)), T.ToTensor()]
        aug  = [T.RandomHorizontalFlip()] if augment else []
        self.tf = T.Compose(aug+base)
    def __len__(self): return len(self.paths)
    def __getitem__(self, idx):
        p = self.paths[idx]
        img = self.tf(Image.open(p).convert('RGB'))
        t = torch.zeros((5,self.S,self.S))
        txt=p.replace('.jpg','.txt')
        if os.path.exists(txt):
            for l in open(txt):
                _,x,y,w,h=map(float,l.split())
                gx,gy=x*self.S,y*self.S; gi,gj=int(gx),int(gy)
                t[0,gj,gi]=gx-gi; t[1,gj,gi]=gy-gj; t[2,gj,gi]=w*self.S; t[3,gj,gi]=h*self.S; t[4,gj,gi]=1.
        return img,t

def collate(b):
    a,b=zip(*b); return torch.stack(a),torch.stack(b)

# =========================
# 3. MODEL (lighter)
# =========================
class TinyYOLO(nn.Module):
    def __init__(self,S=13):
        super().__init__()
        def C(i,o): return nn.Sequential(nn.Conv2d(i,o,3,1,1), nn.BatchNorm2d(o), nn.LeakyReLU(0.1))
        self.back = nn.Sequential(
            C(3,16), nn.MaxPool2d(2),
            C(16,32), nn.MaxPool2d(2),
            C(32,64), nn.MaxPool2d(2),
            C(64,128),nn.MaxPool2d(2),
            C(128,256),nn.MaxPool2d(2)
        )
        self.pred = nn.Conv2d(256,5,1)
        self.S=S
    def forward(self,x): return self.pred(self.back(x))

# =========================
# 4. LOSS
# =========================
L1=nn.SmoothL1Loss()

def loss_fn(p,t,λc=5,λo=1,λn=0.5,pos_w=15.):
    obj=t[:,4:5]>0; bce=nn.BCEWithLogitsLoss(pos_weight=torch.tensor([pos_w],device=p.device))
    lo=bce(p[:,4:5][obj],torch.ones_like(p[:,4:5][obj])) if obj.any() else torch.tensor(0.,device=p.device)
    lno=bce(p[:,4:5][~obj],torch.zeros_like(p[:,4:5][~obj]))
    lb=L1(p[:,:4][obj.expand_as(p[:,:4])],t[:,:4][obj.expand_as(t[:,:4])]) if obj.any() else torch.tensor(0.,device=p.device)
    return λc*lb+λo*lo+λn*lno

# =========================
# 5. TRAIN
# =========================
device=torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model=TinyYOLO().to(device)
opt=optim.Adam(model.parameters(),lr=1e-3)
train_dl=DataLoader(FishDataset(train_imgs,augment=True),batch_size=4,shuffle=True,collate_fn=collate)
val_dl=DataLoader(FishDataset(val_imgs),batch_size=4,collate_fn=collate)

epochs=10
for ep in range(1,epochs+1):
    model.train();run=0
    for x,y in train_dl:
        x,y=x.to(device),y.to(device)
        out=model(x); loss=loss_fn(out,y)
        opt.zero_grad(); loss.backward(); opt.step(); run+=loss.item()
    torch.cuda.empty_cache()
    print(f"E{ep}/{epochs} loss {run/len(train_dl):.4f}")

# =========================
# 6. VERY ROUGH VAL (counts any pred as TP)
# =========================
model.eval(); tp=len(val_imgs); fp=fn=0
print(f"Dummy Precision=1.0 Recall=1.0 (placeholder)")


Resuming download from 128974848 bytes (229544954 bytes left)...
Resuming download from https://www.kaggle.com/api/v1/datasets/download/zehraatlgan/fish-detection?dataset_version_number=1 (128974848/358519802) bytes left.


100%|██████████| 342M/342M [00:10<00:00, 21.1MB/s]

Extracting files...


Using data.yaml: /root/.cache/kagglehub/datasets/zehraatlgan/fish-detection/versions/1/data.yaml
Train dir: /root/.cache/kagglehub/datasets/zehraatlgan/fish-detection/versions/1/train/images
Val dir  : /root/.cache/kagglehub/datasets/zehraatlgan/fish-detection/versions/1/valid/images
E1/10 loss 0.0837
E2/10 loss 0.0039
E3/10 loss 0.0018
E4/10 loss 0.0012
E5/10 loss 0.0010
E6/10 loss 0.0008
E7/10 loss 0.0006
E8/10 loss 0.0005
E9/10 loss 0.0004
E10/10 loss 0.0004
Dummy Precision=1.0 Recall=1.0 (placeholder)


In [3]:
import os, glob, yaml, torch, kagglehub
import torch.nn as nn, torch.optim as optim
from PIL import Image
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as T
from torchvision.ops import box_iou

# =============== 1. DATA LOADING ===============
print("⬇️  Downloading dataset (кешируется)…")
root = kagglehub.dataset_download("zehraatlgan/fish-detection")
yaml_path = os.path.join(root, "data.yaml")
print("Using data.yaml:", yaml_path)

with open(yaml_path) as f:
    cfg = yaml.safe_load(f)
train_dir = os.path.normpath(os.path.join(root, cfg["train"].lstrip("../")))
val_dir   = os.path.normpath(os.path.join(root, cfg["val"].lstrip("../")))
print("Train dir:", train_dir)
print("Val dir  :", val_dir)

train_imgs = sorted(glob.glob(os.path.join(train_dir, "*.jpg")))[:256]
val_imgs   = sorted(glob.glob(os.path.join(val_dir,   "*.jpg")))[:64]
assert train_imgs and val_imgs, "Images not found!"

IMG_SIZE = 416        # ↓ можно 320, если мало GPU-памяти
GRID_S   = 13         # 416 / 32 ≈ 13

# =============== 2. DATASET ===============
class FishDataset(Dataset):
    def __init__(self, paths, S=GRID_S, augment=False):
        self.paths, self.S = paths, S
        base = [T.Resize((IMG_SIZE, IMG_SIZE)), T.ToTensor()]
        aug  = [T.ColorJitter(0.2,0.2,0.2,0.1),
                T.RandomHorizontalFlip()] if augment else []
        self.tf = T.Compose(aug + base)
    def __len__(self): return len(self.paths)
    def __getitem__(self, i):
        p = self.paths[i]
        img = self.tf(Image.open(p).convert("RGB"))
        t = torch.zeros((5, self.S, self.S))
        lab = p.replace(".jpg", ".txt")
        if os.path.exists(lab):
            for line in open(lab):
                _,x,y,w,h = map(float, line.split())
                gx, gy = x*self.S, y*self.S
                gi, gj = int(gx), int(gy)
                if gi>=self.S or gj>=self.S: continue
                t[0,gj,gi] = gx-gi;  t[1,gj,gi] = gy-gj
                t[2,gj,gi] = w*self.S; t[3,gj,gi] = h*self.S
                t[4,gj,gi] = 1.
        return img, t

def collate(b): imgs,tg = zip(*b); return torch.stack(imgs), torch.stack(tg)

# =============== 3. MODEL ===============
class CBL(nn.Sequential):
    def __init__(self, c_in, c_out, k=3, s=1):
        super().__init__(
            nn.Conv2d(c_in, c_out, k, s, k//2, bias=False),
            nn.BatchNorm2d(c_out),
            nn.LeakyReLU(0.1, inplace=True)
        )

class TinyYOLO(nn.Module):
    def __init__(self, S=GRID_S):
        super().__init__()
        self.back = nn.Sequential(
            CBL(3,16),  nn.MaxPool2d(2),
            CBL(16,32), nn.MaxPool2d(2),
            CBL(32,64), nn.MaxPool2d(2),
            CBL(64,128),nn.MaxPool2d(2),
            CBL(128,256),nn.MaxPool2d(2),
            CBL(256,256)
        )
        self.pred = nn.Conv2d(256, 5, 1)   # tx,ty,tw,th,conf
    def forward(self,x): return self.pred(self.back(x))

# =============== 4. LOSS ===============
smooth_l1 = nn.SmoothL1Loss(reduction='mean')
def yolo_loss(p, t, λc=5, λo=1, λn=0.5, pos_w=15.):
    obj = t[:,4:5] > 0
    bce = nn.BCEWithLogitsLoss(pos_weight=torch.tensor([pos_w], device=p.device))
    lo  = bce(p[:,4:5][obj], torch.ones_like(p[:,4:5][obj])) if obj.any() else torch.tensor(0., device=p.device)
    lno = bce(p[:,4:5][~obj], torch.zeros_like(p[:,4:5][~obj]))
    lb  = smooth_l1(p[:,:4][obj.expand_as(p[:,:4])], t[:,:4][obj.expand_as(t[:,:4])]) if obj.any() else torch.tensor(0.,device=p.device)
    return λc*lb + λo*lo + λn*lno

# =============== 5. TRAINING ===============
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model  = TinyYOLO().to(device)
opt    = optim.AdamW(model.parameters(), lr=1e-3, weight_decay=1e-4)

train_dl = DataLoader(FishDataset(train_imgs, augment=True), batch_size=8, shuffle=True,  collate_fn=collate)
val_dl   = DataLoader(FishDataset(val_imgs),               batch_size=8, shuffle=False, collate_fn=collate)

print("🚀 Training…")
for epoch in range(1, 21):
    model.train(); total=0
    for imgs,tg in train_dl:
        imgs, tg = imgs.to(device), tg.to(device)
        loss = yolo_loss(model(imgs), tg)
        opt.zero_grad(); loss.backward(); opt.step()
        total += loss.item()
    torch.cuda.empty_cache()
    print(f"Epoch {epoch:2d}/20 | loss {total/len(train_dl):.4f}")

# =============== 6. VALIDATION ===============
print("🔍 Evaluating…")
model.eval(); tp=fp=fn=0
cell = IMG_SIZE / GRID_S
with torch.no_grad():
    for imgs, tg in val_dl:
        preds = model(imgs.to(device)).cpu()
        for b in range(preds.size(0)):
            p = preds[b]; conf = torch.sigmoid(p[4])
            idx = (conf>0.25).nonzero(as_tuple=False)
            pred_boxes = []
            for gj,gi in idx:
                tx,ty,tw,th = p[0,gj,gi], p[1,gj,gi], p[2,gj,gi], p[3,gj,gi]
                cx = (gi+torch.sigmoid(tx))*cell
                cy = (gj+torch.sigmoid(ty))*cell
                w  = torch.exp(tw).clamp(max=5)*cell
                h  = torch.exp(th).clamp(max=5)*cell
                pred_boxes.append([cx-w/2, cy-h/2, cx+w/2, cy+h/2])
            pred_boxes = torch.tensor(pred_boxes) if pred_boxes else torch.zeros((0,4))

            gt = []
            for gj,gi in (tg[b,4]>0).nonzero(as_tuple=False):
                tx,ty,tw,th = tg[b,0,gj,gi], tg[b,1,gj,gi], tg[b,2,gj,gi], tg[b,3,gj,gi]
                cx=(gi+tx)*cell; cy=(gj+ty)*cell; w=tw*cell; h=th*cell
                gt.append([cx-w/2, cy-h/2, cx+w/2, cy+h/2])
            gt = torch.tensor(gt) if gt else torch.zeros((0,4))

            if pred_boxes.numel()==0: fn += len(gt); continue
            if gt.numel()==0: fp += len(pred_boxes); continue
            iou = box_iou(pred_boxes, gt)
            m,_ = iou.max(1)
            tp += (m>=0.5).sum().item()
            fp += (m<0.5).sum().item()
            fn += gt.size(0) - (m>=0.5).sum().item()

prec = tp/(tp+fp+1e-8); rec = tp/(tp+fn+1e-8)
f1   = 2*prec*rec/(prec+rec+1e-8)
print(f"✅ Precision={prec:.3f} Recall={rec:.3f} F1={f1:.3f}")


⬇️  Downloading dataset (кешируется)…
Using data.yaml: /kaggle/input/fish-detection/data.yaml
Train dir: /kaggle/input/fish-detection/train/images
Val dir  : /kaggle/input/fish-detection/valid/images
🚀 Training…
Epoch  1/20 | loss 0.1487
Epoch  2/20 | loss 0.0208
Epoch  3/20 | loss 0.0086
Epoch  4/20 | loss 0.0051
Epoch  5/20 | loss 0.0035
Epoch  6/20 | loss 0.0025
Epoch  7/20 | loss 0.0020
Epoch  8/20 | loss 0.0016
Epoch  9/20 | loss 0.0013
Epoch 10/20 | loss 0.0011
Epoch 11/20 | loss 0.0009
Epoch 12/20 | loss 0.0008
Epoch 13/20 | loss 0.0007
Epoch 14/20 | loss 0.0006
Epoch 15/20 | loss 0.0005
Epoch 16/20 | loss 0.0005
Epoch 17/20 | loss 0.0004
Epoch 18/20 | loss 0.0004
Epoch 19/20 | loss 0.0003
Epoch 20/20 | loss 0.0003
🔍 Evaluating…
✅ Precision=0.000 Recall=0.000 F1=0.000
